File: python/grade-school/grade_school.py
from collections import defaultdict
from functools import reduce


class School(object):
    def __init__(self):
        self.directory = {}
        self.was_added = []

    def add_student(self, name, grade):
        if name in self.directory:
            self.was_added.append(False)
        else:
            self.directory[name] = grade
            self.was_added.append(True)

    def roster(self):
        roster = []
        for grade in sorted(set(self.directory.values())):
            roster += self.grade(grade)
        return roster

    def grade(self, grade_number):
        names = [
            name for name, grade in self.directory.items()
                 if grade == grade_number
        ]
        return sorted(names)

    def added(self):
        return self.was_added

File: nim/grade-school/grade_school.nim
from algorithm import sorted, sortedByIt
from sequtils  import filterIt, mapIt

type
  Student = tuple
    name: string
    grade: int

  School* = object
    students*: seq[Student]


# list of student names, sorted first by grade then by name
proc roster*(school: School): seq[string] =
  school
    .students
    .sortedByIt((it.grade, it.name))
    .mapIt(it.name)


# list of students in the requested grade, sorted by name
proc grade*(school: School, grade: int): seq[string] =
  school
    .students
    .filterIt(it.grade == grade)
    .mapIt(it.name)
    .sorted
